### ISSUE TITLE
What is the intuition behind using TimeDistributedLayer?

### ISSUE THREAD

**Issue by Abhijit-2592** [yhenon/keras-frcnn#208]
@yhenon  Hey what is your intuition behind using TimeDistributed(Dense()) instead of just Dense()? Does it affect the detection performance of the network?

**Comment by thomasjanssens** [comment]
I also spent a long time looking at this. If I understand it correctly, _TimeDistributed(L)(T)_ applies layer _L_ to each slice of your tensor _T_ along axis 1. This is used in the network to pass along all bounding boxes in one tensor with dimensionality [1,number_of_boxes,....] to the roi pooling/classifier branch of the net. TimeDistributed then slices this big tensor along the second axis (axis 1) into the individual bounding boxes so that the classifier is applied to each one individually.

I guess the "TimeDistributed" name stems from the fact that this was probably created to slice things along a time axis, but it's a quite misleading name, in my opinion, since it can be used in a non-temporal context too.

**Comment by Abhijit-2592** [comment]
Mmm why second axis? Am still confused

**Comment by thomasjanssens** [comment]
I'm not sure why TimeDistributed as defined in the Keras API works on specifically the second axis, but I believe in the code here a "dummy" axis of length 1 is added in the front of the tensor as a hack to make the roi axis (which we want to split along) the second one. It would be easier if you could specify the axis you want to split along, but I guess you can't and this is a trick to get around that limitation.

**Comment by Tbarowski** [comment]
So this is more or less a kind of batch management where you don't need to providing an input for each sample?

**Comment by thomasjanssens** [comment]
Indeed, it performs the same operation X times on all [A,B,C]-shaped slices of a tensor shaped [1,X,A,B,C].
